frontend/src/components/widgets/DataFrame/DataFrameCells.tsx

```diff
@@ -0,0 +1,280 @@
+    return "smart"
+  }
+
+  return "default"
+}
+
+/**
+ * Returns a glide-data-grid compatible cell object based on the
+ * cell data from the quiver object. Different types of data will
+ * result in different cell types.
+ *
+ * @param cellTemplate: the empty cell template from the column.
+ * @param quiverCell: a dataframe cell object from Quiver.
+ * @param cssStyles: optional css styles to apply on the cell.
+ *
+ * @return a GridCell object that can be used by glide-data-grid.
+ */
+export function fillCellTemplate(
+  cellTemplate: GridCell,
+  quiverCell: DataFrameCell,
+  cssStyles: string | undefined = undefined
+): GridCell {
+  let cellKind = cellTemplate.kind
+  if (cellTemplate.kind === GridCellKind.Custom) {
+    cellKind = (cellTemplate.data as any)?.kind
+
+    if (!cellKind) {
+      throw new Error(`Unable to determine cell type for custom cell.`)
+    }
+  }
+
+  if (cssStyles && quiverCell.cssId) {
+    const themeOverride = {}
+
+    // Extract and apply the font color
+    const fontColor = extractCssProperty(quiverCell.cssId, "color", cssStyles)
+    if (fontColor) {
+      ;(themeOverride as GlideTheme).textDark = fontColor
+    }
+
+    // Extract and apply the background color
+    const backgroundColor = extractCssProperty(
+      quiverCell.cssId,
+      "background-color",
+      cssStyles
+    )
+    if (backgroundColor) {
+      ;(themeOverride as GlideTheme).bgCell = backgroundColor
+    }
+
+    if (themeOverride) {
+      // Apply the background and font color in the theme override
+      cellTemplate = {
+        ...cellTemplate,
+        themeOverride,
+      }
+    }
+  }
+
+  if (cellKind === GridCellKind.Text) {
+    const formattedContents = getDisplayContent(quiverCell)
+    return {
+      ...cellTemplate,
+      data:
+        typeof quiverCell.content === "string" ||
+        quiverCell.content === undefined ||
+        quiverCell.content === null
+          ? quiverCell.content
+          : formattedContents,
+      displayData: formattedContents,
+    } as TextCell
+  }
+
+  if (cellKind === GridCellKind.Number) {
+    const formattedContents = getDisplayContent(quiverCell)
+    let cellData = quiverCell.content
+
+    if (cellData instanceof Int32Array) {
+      // int values need to be extracted this way:
+      // eslint-disable-next-line prefer-destructuring
+      cellData = (cellData as Int32Array)[0]
+    }
+
+    return {
+      ...cellTemplate,
+      data:
+        cellData !== undefined && cellData !== null
+          ? Number(cellData)
+          : undefined,
+      displayData: formattedContents,
+    } as NumberCell
+  }
+
+  if (cellKind === GridCellKind.Boolean) {
+    return {
+      ...cellTemplate,
+      data: quiverCell.content as boolean,
+    } as BooleanCell
+  }
+
+  if (cellKind === GridCellKind.Bubble) {
+    return {
+      ...cellTemplate,
+      data:
+        quiverCell.content !== undefined && quiverCell.content !== null
```

Comment:
Seems like the undefined and null check could be a separate helper function or something, just to simplify and be used in multiple places.

Reply:
I added a helper to `utils.tsx`. There we already had `notNull` and `notUndefined` functions.